### PR TITLE
Fix overscaling during app upgrade

### PR DESCRIPTION
**Pull request by zen-dog**
SI `test_app_update_rollback` caught a bug where `TaskReplaceActor` would overscale during app upgrades due to the recent introduction of instance goals. This commit fixes this bug.

Fixes: MARATHON-8537

### CODE REVIEW

File: src/main/scala/mesosphere/marathon/core/deployment/impl/DeploymentManagerActor.scala [hunk 1]
```diff
@@ -254,7 +254,7 @@ class DeploymentManagerActor(
     async {
       await(Future.sequence(cancellations))
 
-      logger.info(s"Conflicting deployments ${toCancel.map(_.plan.id)} for deployment ${plan.id} have been canceled")
+      logger.info(s"Conflicting deployments ${conflicts.map(_.plan.id)} for deployment ${plan.id} have been canceled")
```
Comment (zen-dog) [inline src/main/scala/mesosphere/marathon/core/deployment/impl/DeploymentManagerActor.scala]: unrelated logging improvement

File: src/main/scala/mesosphere/marathon/core/deployment/impl/TaskReplaceActor.scala [hunk 2]
```diff
@@ -35,12 +35,12 @@ class TaskReplaceActor(
...
   // compute all values ====================================================================================
 
-  // All running instances of this app
+  // All existing instances of this app independent of the goal.
   //
   // Killed resident tasks are not expunged from the instances list. Ignore
   // them. LaunchQueue takes care of launching instances against reservations
   // first
-  val currentInstances = instanceTracker.specInstancesSync(runSpec.id).filter(_.state.goal == Goal.Running)
+  val currentInstances = instanceTracker.specInstancesSync(runSpec.id)
```
Comment (zen-dog) [inline src/main/scala/mesosphere/marathon/core/deployment/impl/TaskReplaceActor.scala]: Bug nr. 1: we would only take `Running` instances into consideration, however, during an upgrade there would be old instances in the process of being killed with a `Goal.Decommissioned`

File: src/main/scala/mesosphere/marathon/core/deployment/impl/TaskReplaceActor.scala [hunk 3]
```diff
@@ -111,12 +111,13 @@ class TaskReplaceActor(
...
   def replaceBehavior: Receive = {
-    // New instance failed to start, restart it
-    case InstanceChanged(id, runSpecVersion, `pathId`, condition, Instance(instanceId, Some(agentInfo), state, tasksMap, runSpec, reservation)) if newInstanceIds(id) && considerTerminal(condition) =>
-      logger.warn(s"New instance $id is terminal on agent ${agentInfo.agentId} during app $pathId restart: $condition reservation: $reservation")
-      instanceTerminated(id)
-      instancesStarted -= 1
-      launchInstances().pipeTo(self)
+    // New instance task failed: instance goal is still Running so nothing to do here, the task will be rescheduled automatically
+    case InstanceChanged(id, runSpecVersion, `pathId`, condition, Instance(instanceId, Some(agentInfo), state, tasksMap, runSpec, reservation)) if newInstanceIds(id) && considerTerminal(condition) && state.goal == Goal.Running =>
+      logger.warn(s"New $id is terminal ($condition) on agent ${agentInfo.agentId} during app $pathId restart: $condition reservation: $reservation. Waiting for the task to restart...")
```
Comment (zen-dog) [inline src/main/scala/mesosphere/marathon/core/deployment/impl/TaskReplaceActor.scala]: Bug nr. 2: If a new instance task would fail, we would create a new one (`launchInstances()` call on the left side). This is not necessary anymore since the instance is already new and the task will be rescheduled by the `TaskLauncherActor`

File: src/main/scala/mesosphere/marathon/core/deployment/impl/TaskReplaceActor.scala [hunk 4]
```diff
@@ -111,12 +111,13 @@ class TaskReplaceActor(
...
+
+    // New instance task failed: instance goal is *NOT* Running?! This should *NOT* happen and means, someone is interfering with current deployment!
+    case InstanceChanged(id, runSpecVersion, `pathId`, condition, Instance(instanceId, Some(agentInfo), state, tasksMap, runSpec, reservation)) if newInstanceIds(id) && considerTerminal(condition) && state.goal != Goal.Running =>
+      logger.error(s"New $id is terminal ($condition) on agent ${agentInfo.agentId} during app $pathId restart (reservation: $reservation) and the goal (${state.goal}) is *NOT* Running! This means that someone is interfering with current deployment!")
```
Comment (zen-dog) [inline src/main/scala/mesosphere/marathon/core/deployment/impl/TaskReplaceActor.scala]: This is more of a sanity check: basically, we don't expect anybody to interfere with the deployment (`MarathonSchedulerActor` lock should prevent this). Not sure what to do here since only one orchestrator logic should be responsible for setting instance goal and this behavior is considered a bug now. The deployment will be basically "stuck" (and has to be canceled/stopped by the user should this BUG surface)

File: src/main/scala/mesosphere/marathon/core/deployment/impl/TaskReplaceActor.scala (hunk 2 above)
Comment (jeschkies) [reply src/main/scala/mesosphere/marathon/core/deployment/impl/TaskReplaceActor.scala]: But these should not be kill, right? As far as I know the `currentInstances` also select the old instances that should receive a kill.

File: src/main/scala/mesosphere/marathon/core/deployment/impl/TaskReplaceActor.scala (hunk 4 above)
Comment (jeschkies) [reply src/main/scala/mesosphere/marathon/core/deployment/impl/TaskReplaceActor.scala]: Why not nest this receive and the one above? It is hard to follow all `if`-statements after the extractor. What do you think about defining a new extractor?

```scala
case InstanceChanged(id, runSpecVersion, `pathId`, condition, Instance(instanceId, Some(agentInfo), State(Running, Terminal), tasksMap, runSpec, reservation)) if newInstanceIds(id)
```

File: src/main/scala/mesosphere/marathon/core/deployment/impl/TaskReplaceActor.scala (hunk 2 above)
Comment (jeschkies) [reply src/main/scala/mesosphere/marathon/core/deployment/impl/TaskReplaceActor.scala]: Come to think of it I think we should update this instance set on each update and then call killing, launching and finished check. That would simple the logic quite a bit. It would break some unit tests and require a little refactoring in the starting behaviour.

Comment (zen-dog) [reply src/main/scala/mesosphere/marathon/core/deployment/impl/TaskReplaceActor.scala]: Yeah, should be a separate effort
